crates/bevy_asset/src/http_source.rs

```diff
@@ -0,0 +1,242 @@
+    use crate::io::wasm::HttpWasmAssetReader;
+
+    HttpWasmAssetReader::new("")
+        .fetch_bytes(path)
+        .await
+        .map(|r| Box::new(r) as Box<dyn Reader>)
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+async fn get<'a>(path: PathBuf) -> Result<Box<dyn Reader>, AssetReaderError> {
+    use crate::io::VecReader;
+    use std::io;
+
+    let str_path = path.to_str().ok_or_else(|| {
+        AssetReaderError::Io(
+            io::Error::new(
+                io::ErrorKind::Other,
+                format!("non-utf8 path: {}", path.display()),
+            )
+            .into(),
+        )
+    })?;
+
+    if let Some(data) = http_asset_cache::try_load_from_cache(str_path)? {
+        return Ok(Box::new(VecReader::new(data)));
+    }
+
+    match ureq::get(str_path).call() {
+        Ok(response) => {
+            let mut reader = response.into_reader();
+            let mut buffer = Vec::new();
+            reader.read_to_end(&mut buffer)?;
+
+            http_asset_cache::save_to_cache(str_path, &buffer)?;
+
+            Ok(Box::new(VecReader::new(buffer)))
+        }
+        // ureq considers all >=400 status codes as errors
+        Err(ureq::Error::Status(code, _response)) => {
+            if code == 404 {
+                Err(AssetReaderError::NotFound(path))
+            } else {
+                Err(AssetReaderError::HttpError(code))
+            }
+        }
+        Err(ureq::Error::Transport(err)) => Err(AssetReaderError::Io(
+            io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "unexpected error while loading asset {}: {}",
+                    path.display(),
+                    err
+                ),
+            )
+            .into(),
+        )),
+    }
+}
+
+impl AssetReader for HttpSourceAssetReader {
+    fn read<'a>(
+        &'a self,
+        path: &'a Path,
+    ) -> impl ConditionalSendFuture<Output = Result<Box<dyn Reader>, AssetReaderError>> {
+        get(self.make_uri(path))
+    }
+
+    async fn read_meta<'a>(&'a self, path: &'a Path) -> Result<Box<dyn Reader>, AssetReaderError> {
+        match self.make_meta_uri(path) {
+            Some(uri) => get(uri).await,
+            None => Err(AssetReaderError::NotFound(
+                "source path has no extension".into(),
+            )),
+        }
+    }
+
+    async fn is_directory<'a>(&'a self, _path: &'a Path) -> Result<bool, AssetReaderError> {
+        Ok(false)
+    }
+
+    async fn read_directory<'a>(
+        &'a self,
+        path: &'a Path,
+    ) -> Result<Box<PathStream>, AssetReaderError> {
+        Err(AssetReaderError::NotFound(self.make_uri(path)))
+    }
+}
+
+/// A naive implementation of an HTTP asset cache that never invalidates.
+/// `ureq` currently does not support caching, so this is a simple workaround.
+/// It should eventually be replaced by `http-cache` or similar, see [tracking issue](https://github.com/06chaynes/http-cache/issues/91)
+mod http_asset_cache {
```

Comment:
I'm a bit sceptical about the caching implementation, mostly because I am biased and my game's assets are small enough not to need caching and I don't want my game to create any unnecessary folders. But also because the default behaviour creating a hidden folder might mean some users may not even realise the assets are being cached, especially since it isn't documented on any high level docs of this feature.

It's clearly useful but IMO caching should be disabled by default as it's something that I think people will reach for if they need it. I don't know if its worth blocking on this but I think it deservers a bit more discussion at least😄

Reply:
Yep good point ive exposed a `http_source_cache` top level feature.